internal/app/client/conformance_client.go

```diff
@@ -218,6 +221,118 @@ func (i *invoker) clientStream(
 	}, nil
 }
 
+func (i *invoker) bidiStream(
+	ctx context.Context,
+	req *v1alpha1.ClientCompatRequest,
+) (*v1alpha1.ClientResponseResult, error) {
+	stream := i.client.BidiStream(ctx)
+
+	// Add the specified request headers to the request
+	app.AddHeaders(req.RequestHeaders, stream.RequestHeader())
+
+	firstSend := true
+	fullDuplex := false
+
+	var protoErr *v1alpha1.Error
+	var headers []*v1alpha1.Header
+	var trailers []*v1alpha1.Header
+	payloads := make([]*v1alpha1.ConformancePayload, 0, 1)
+
+	for _, msg := range req.RequestMessages {
+		if err := ctx.Err(); err != nil {
+			// If an error was returned, convert it to a proto Error
+			protoErr = app.ConvertErrorToProtoError(err)
+			break
+		}
+		bsr := &v1alpha1.BidiStreamRequest{}
+		if err := msg.UnmarshalTo(bsr); err != nil {
+			return nil, err
+		}
+		if firstSend {
+			fullDuplex = bsr.FullDuplex
+		}
+		if err := stream.Send(bsr); err != nil && errors.Is(err, io.EOF) {
+			// TODO - Is this correct? Docs say the returned error will wrap EOF
+			// and clients should call Receive to get the error so assuming this error
+			// is the unwrapped error from the send call?
+			// Call receive to get the error
+			_, err := stream.Receive()
+			if err != nil {
+				// If an error was returned, convert it
+				// to a proto Error
+				protoErr = app.ConvertErrorToProtoError(err)
+				// Break the send loop
+				break
+			}
+		}
+		if fullDuplex {
+			// If this is a full duplex stream, receive a response for each request
+			msg, err := stream.Receive()
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					// If an error was returned that is not an EOF, convert it
+					// to a proto Error
+					protoErr = app.ConvertErrorToProtoError(err)
+				}
+				// Reads are done either because we received an error or an EOF
+				// In either case, break the outer loop
+				break
+			}
+			// If the call was successful, get the returned payloads
+			payloads = append(payloads, msg.Payload)
+		}
+	}
+
+	// If protoErr is nil, then the send loop terminated successfully
+	if protoErr == nil {
+		// Close the send side of the stream
+		if err := stream.CloseRequest(); err != nil {
+			protoErr = app.ConvertErrorToProtoError(err)
+		}
+	}
+
+	// If there were no errors encountered yet, then receive any additional responses
+	if protoErr == nil {
```

Comment:
Feels gross to check `protoErr` here again, but this is basically checking "the send loop completed successfully, but closing the send side of the stream failed, so don't try to receive any additional responses"

Reply:
Instead of `if protoErr == nil`, I think it would be better to invert the condition and early-return. So `if protoErr != nil`, we're done.

To do that w/out lots of boiler-plate, you could use a deferred function that does all of the stuff at the end that is needed using named returns.

So if the signature used named returns like so:
```go
func (i *invoker) bidiStream(
	ctx context.Context,
	req *v1alpha1.ClientCompatRequest,
) (result *v1alpha1.ClientResponseResult, err error) {
	result = &v1alpha1.ClientResponseResult{}
```
Then you can directly mutate the result as you collect payloads, etc (e.g. use `result.Payloads = append(result.Payloads, msg.Payload)`).

So early return could be something like this:
```go
if protoErr != nil {
	result.Error = protoErr
	return result, nil
}
```

Finally, to catch the headers and trailers at the end, you could use a defer function installed at the beginning:
```go
defer func() {
	if result != nil {
		result.Headers = app.ConvertToProtoHeader(stream.ResponseHeader())
		result.Trailers = app.ConvertToProtoHeader(stream.ResponseTrailer())
	}
}()
```